### ISSUE TITLE
Input Lag in Hatsune Miku Project Diva 2nd

### ISSUE THREAD

**Issue by GamerzHell9137** [hrydgard/ppsspp#4020]
It seems that the most important part of a Rhythm games, the inputs, are delaying in Project Diva 2nd.
It works normal on JPCSP but on PPSSPP it doesn't.
And just want to say that Project Diva Extend doesn't suffer of that problem.

**Comment by PierBover** [comment]
I'm playing the three Patapon games and I get a lag of about 300-400 ms, which makes it impossible to play. I've tried both Android and Windows, bluetooth controllers, wired controllers, and keyboards, with the same result. I have tried changing all options related to sound, input, graphics, etc, with no success.

I don't have noticed the lag on other games, but it may be because timing is not so critical...

Im on 1.0.1

**Comment by Wowfunhappy** [comment]
Some good news, as of 1.4.2-198, this game no longer appears to have any sync/timing issues.

I'm using Mac build, on a machine that previously had lag issues when I last tried (around ppsspp v1 if I remember correctly). Audio latency is set to low in settings.

If you've had issues running the Project Diva games in the past, you may want to try again!

**Comment by GamerzHell9137** [comment]
@Wowfunhappy its same, there's been no changes.

**Comment by Wowfunhappy** [comment]
Well, that's odd, because it definitely improved for me on the same machine—the game was unplayable before. I suppose something else in my system could theoretically have changed though.

**Comment by GamerzHell9137** [comment]
Improved? Yes. Like it should be? No.

**Comment by leopard20** [comment]
@unknownbrackets @hrydgard 
It works fine on Android. I've completed the game without any problem. 
However, I did get input lag when I connected headphones. But it wasn't a PPSSPP bug. (I fixed it by simply clearing my RAM cache (using memory optimization tools). Which means it was because of a background app.

**Comment by unknownbrackets** [comment]
Just a note, as seen in #11813, the real clock sync setting was buggy before but should work correctly now.  Note that as per #10195, if you're using OpenGL of Vulkan, you're likely to have some lag in the screen updating versus what's happening in the game.

It might be worth trying this now with the force sync setting enabled and using D3D9 or D3D11.  Note that forcing clock sync is slower, and works best on Linux/Mac (Windows will have ~2ms input lag because of how its timing functions work, but they work accurately on Linux/Mac/Android - of course Android doesn't have D3D.)

-[Unknown]

**Comment by GamerzHell9137** [comment]
@unknownbrackets so what's the best settings on Windows now to get as less as possible input lag?

**Comment by GamerzHell9137** [comment]
@unknownbrackets Other than Force Real Sync and DX9/11, any other option that can improve the input lag on Windows?

**Comment by unknownbrackets** [comment]
Well, the lower quality settings are presumably better.  Make sure you're not using locked CPU speed or any audio hacks.  In theory, non-buffered rendering might have slightly less lag (I'm not sure.)

Just to mention, it's entirely possible that the physical PSP has some input lag (the LCD was slow) and that these games account for it by applying a negative offset to controller input.  In some ways, this could explain the earlier comments about multithreading improving the situation.

If the game does that (once #10195 is fixed, which most likely? is slower than the PSP would be), the problem could be that we're BETTER than the PSP, and so the offset the game uses goes in the wrong direction and makes it think your presses are too early.

That means there might be a "goldilocks zone" for input lag.  "My score was lower than expected" is a binary measurement, so it will be difficult to zero in on the +/- of that zone if there's no way to say "I average too late" or "I average too early" compared to playing on a PSP.

-[Unknown]

**Comment by GamerzHell9137** [comment]
@unknownbrackets Yeah, non buffered makes the input better but in general it feels almost the same. A bit better but yeah, still feels "weird". Don't have a PSP but its funny how Vita and PPSSPP on PC feel different with the input.

About input offsets, is it possible to implement something like that in PPSSPP? The newer Diva games have the option on PS3 and PS Vita which compensates for LCD lag. Still idk if that would fix it, PPSSPP always had this feeling of "you press the button but the game reacts later to it".

And I remember before Megpoid the Music # had less input lag with higher refresh rate monitors (played on 75 Hz), i guess the game reacts different than Diva because it didn't help Diva with input.

**Comment by unknownbrackets** [comment]
The trouble is that it's game logic applying the offset.

Retroarch implemented something for NES games that basically uses savestates every single frame.  But NES games had about 4KB of RAM, the PSP has 8000 times as much.  For a PSP, saving state every frame is too slow and would CAUSE input lag, not solve it.  RAM has only gotten about ~30 times faster maybe since the NES; size gains have outpaced speed gains.

It might be possible for someone to create a cwcheat or hack to adjust the input lag offset the game logic uses - assuming it even exists.

Basically it would have to change the logic so that the "right time" to press a button and get a score is +input lag ms after it's actually displayed.  PPSSPP has to run the game logic to generate graphics, so it only has the ability to move everything together (without the save state thing I mentioned above, which really is a smart idea, but as I said not practical for a PSP game.)

-[Unknown]

**Comment by GamerzHell9137** [comment]
That sucks, welp i guess we gotta wait for that #10195 fix

**Comment by LunaMoo** [comment]
Imo input lag offset wouldn't cause people with this issue to get better at the game, I would rather say that the "Input lag" imo should be even less of a problem in *rhytm games* than any other genre as even without game offsetting it, such games are awfully predictable and any player could offset it himself or starting playing on an emulator even not feel anything wrong with it.

 I do believe based on ways DrawSyncEatCycles hack noticably affecting some of those games(even fixes #9842) and older issue about patapon 1 fever being hard to keep (#2498 ~ which I believe might still be an issue), that the actual problem is that some key presses are expected to happen at different timing than the rest, I recall we had some thread on forum about Patapon 1(long after the issue about it was closed hence I believe it's still a thing) where I helped someone to track some code in disassembly and from that I recall that fourth beat had different timing than the previous 3 when the fever was up, making it much harder to keep, but since it was weird in same way each time, eventually I couldn't tell if it was really broken or not since I got used to it after a few failures..

 PS Vita also has imperfect PSP emulator, definitely not cycle accurate, despite being an "official emulator" it's likely even less compatible than PPSSPP at this point in time and should never be used as an example of how games should work on the real hardware.

**Comment by leopard20** [comment]
Just wanted to let you know that enabling real clock sync on Android as well significantly improves the lag input (Vulkan and OGL). There's a tiny lag as mentioned, but it's negligible compared to before.
PD Extend still plays best (almost completely lag-free) in all diva games.

**Comment by benderscruffy** [comment]
No feedback/outdated?

**Comment by unknownbrackets** [comment]
I feel like there haven't been many people saying this isn't good for a while, and #10195 already covers "all" Proejct Diva games, which presumably includes Project Diva 2nd.  So I'm going to just close this as a duplicate (even though that was opened more recently.)

-[Unknown]
